synapse/tools/rstorm.py

```diff
@@ -18,17 +27,143 @@ async def cortex():
+        ('--hide-query', {}),
+    ) + s_cmds_cortex.StormCmd._cmd_syntax
+
+    def __init__(self, core, ctx, stormopts=None, opts=None):
+        if opts is None:
+            opts = {}
+
+        s_cmds_cortex.StormCmd.__init__(self, None, **opts)
+
+        self.stormopts = stormopts or {}
+
+        # hide a few mesg types by default
+        for mtype in ('init', 'fini', 'node:edits', 'node:edits:count', 'prov:new'):
+            self.cmdmeths[mtype] = self._silence
+
+        self.core = core
+        self.ctx = ctx
+        self.lines = []
+
+    async def runCmdLine(self, line):
+        opts = self.getCmdOpts(f'storm {line}')
+        return await self.runCmdOpts(opts)
+
+    async def _mockHttp(self, *args, **kwargs):
+        info = {
+            'code': 200,
+            'body': '{}',
+        }
+
+        resp = self.ctx.get('mock-http')
+        if resp:
+            body = resp.get('body')
+
+            if isinstance(body, dict):
+                body = json.dumps(body)
+
+            info = {
+                'code': resp.get('code', 200),
+                'body': body,
+            }
+
+        return s_stormhttp.HttpResp(info)
+
+    def printf(self, mesg, addnl=True, color=None):
+        line = f'    {mesg}'
+        self.lines.append(line)
+        return line
+
+    def _silence(self, mesg, opts):
+        pass
+
+    def _onErr(self, mesg, opts):
+        # raise on err for rst
+        raise s_exc.StormRuntimeError(mesg=mesg)
+
+    async def runCmdOpts(self, opts):
+
+        text = opts.get('query')
+
+        if not opts.get('hide-query'):
+            self.printf(f'> {text}')
+
+        stormopts = copy.deepcopy(self.stormopts)
+
+        stormopts.setdefault('repr', True)
+        stormopts.setdefault('path', opts.get('path', False))
+
+        hide_unknown = True
+
+        # TODO
+        # showtext = opts.get('show')
+        # if showtext is not None:
+        #     stormopts['show'] = showtext.split(',')
+        #
+        # editformat = opts['editformat']
+        # if editformat != 'nodeedits':
+        #     stormopts['editformat'] = editformat
+
+        stormopts['editformat'] = 'nodeedits'
+
+        # Let this raise on any errors
+        with mock.patch('synapse.lib.stormhttp.LibHttp._httpRequest', new=self._mockHttp):
+            async for mesg in self.core.storm(text, opts=stormopts):
+
+                if opts.get('debug'):
+                    self.printf(pprint.pformat(mesg))
+                    continue
+
+                try:
+                    func = self.cmdmeths[mesg[0]]
+                except KeyError:  # pragma: no cover
+                    if hide_unknown:
+                        continue
+                    self.printf(repr(mesg), color=s_cmds_cortex.UNKNOWN_COLOR)
+                else:
+                    func(mesg, opts)
+
+        return '\n'.join(self.lines)
+
+
 async def genStormRst(path, debug=False):
 
     outp = []
     context = {}
 
+    def reqCore():
```

Comment:
We should probably be using an contextlib.AsyncExitStack here to ensure that we have a way to ensure that cortexes made are always fini'd and their associated temp dirs get cleaned up.

Reply:
Moved things around per discussion